src/fedify/mq/gcloud-pubsub-mq.ts

```diff
@@ -0,0 +1,105 @@
+    MessageQueueEnqueueOptions,
+    MessageQueueListenOptions,
+} from '@fedify/fedify';
+import type { Logger } from '@logtape/logtape';
+
+export class GCloudPubSubMessageQueue implements MessageQueue {
+    private pubSubClient: PubSub;
+    private topicIdentifier: string;
+    private subscriptionIdentifier: string;
+    private logger: Logger;
+
+    constructor(
+        pubSubClient: PubSub,
+        topicIdentifier: string,
+        subscriptionIdentifier: string,
+        logger: Logger,
+    ) {
+        this.pubSubClient = pubSubClient;
+        this.topicIdentifier = topicIdentifier;
+        this.subscriptionIdentifier = subscriptionIdentifier;
+        this.logger = logger;
+    }
+
+    async enqueue(
+        message: any,
+        options?: MessageQueueEnqueueOptions,
+    ): Promise<void> {
+        const delay = options?.delay?.total('millisecond') ?? 0;
+
+        this.logger.info(
+            `Enqueuing message [FedifyID: ${message.id}] with delay: ${delay}ms`,
+        );
+
+        if (delay > 0) {
+            await new Promise((resolve) => setTimeout(resolve, delay));
+        }
+
+        try {
+            const messageId = await this.pubSubClient
+                .topic(this.topicIdentifier)
+                .publishMessage({
+                    json: message,
+                    attributes: {
+                        fedifyId: message.id,
+                    },
+                });
+
+            this.logger.info(
+                `Message [FedifyID: ${message.id}] was enqueued [PubSubID: ${messageId}]`,
+            );
+        } catch (error) {
+            this.logger.error(
+                `Failed to enqueue message [FedifyID: ${message.id}]: ${error}`,
+            );
+        }
+    }
+
+    async listen(
+        handler: (message: any) => Promise<void> | void,
+        options: MessageQueueListenOptions = {},
+    ): Promise<void> {
+        const subscription = this.pubSubClient.subscription(
+            this.subscriptionIdentifier,
+        );
+
+        subscription.on('message', async (message: Message) => {
+            const fedifyId = message.attributes.fedifyId ?? 'unknown';
+
+            this.logger.info(
+                `Handling message [FedifyID: ${fedifyId}, PubSubID: ${message.id}]`,
+            );
+
+            try {
+                const json = JSON.parse(message.data.toString());
+
+                await handler(json);
+
+                message.ack();
+
+                this.logger.info(
+                    `Acknowledged message [FedifyID: ${fedifyId}, PubSubID: ${message.id}]`,
+                );
+            } catch (error) {
+                message.nack();
+
+                this.logger.error(
+                    `Failed to handle message [FedifyID: ${fedifyId}, PubSubID: ${message.id}]: ${error}`,
+                );
+            }
+        });
+
+        return await new Promise((resolve) => {
+            options.signal?.addEventListener('abort', () => {
```

Comment:
If signal is not defined, this promise will never resolve - is that expected?

Reply:
I think so yeh, from the docs:

> A Promise object it returns should never resolve unless the given [signal](https://jsr.io/@fedify/fedify@1.3.0-dev.492+450a6939/doc/~/MessageQueueListenOptions.signal) is triggered.

Reply:
Oh my bad, you meant if `signal` is not provided at all. I wonder why that would be the case 🤔 

@dahlia In what cases would `signal` not be provided?

Reply:
The `signal` is usually not provided, unless you're [manually starting the worker](https://fedify.dev/manual/mq#separating-message-processing-from-the-main-process). And yes, this promise is expected to be never resolved unless any abort signal is invoked.

Reply:
What about if the underlying listener is aborted unexpectedly - should the promise resolve then?